File: legacyjava/core/src/main/java/org/javafish/move/MoveList.java
package org.javafish.move;

import org.javafish.bitboard.Bitboard;
import org.javafish.board.BoardState;
import org.javafish.board.Side;
import search.MoveOrdering;
import search.TTEntry;
import search.TranspositionTable;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;

import static org.javafish.eval.PieceSquareTable.MGS;

public class MoveList extends ArrayList<Move> {

    private List<ScoredMove> sortedList;
    public MoveList(){}

//    public MoveList(MoveList halfMoves){
//        super.addAll(halfMoves);
//    }

    public void makeQ(int fromSq, long to){
        int toSq;
        while (to != 0){
            toSq = Bitboard.lsb(to);
            to = Bitboard.extractLsb(to);
            super.add(new Move(fromSq, toSq, Move.QUIET));
        }
    }

    public void makeC(int fromSq, long to){
        int toSq;
        while (to != 0){
            toSq = Bitboard.lsb(to); // TODO inline
            to = Bitboard.extractLsb(to);
            super.add(new Move(fromSq, toSq, Move.CAPTURE));
        }
    }

    public void makeDP(int fromSq, long to){
        int toSq;
        while (to != 0){
            toSq = Bitboard.lsb(to);
            to = Bitboard.extractLsb(to);
            super.add(new Move(fromSq, toSq, Move.DOUBLE_PUSH));
        }
    }

//    public void makePR(int fromSq, long to){
//        int toSq;
//        while (to != 0){
//            toSq = Bitboard.lsb(to);
//            to = Bitboard.extractLsb(to);
//            super.add(new Move(fromSq, toSq, Move.PR_KNIGHT));
//            super.add(new Move(fromSq, toSq, Move.PR_BISHOP));
//            super.add(new Move(fromSq, toSq, Move.PR_ROOK));
//            super.add(new Move(fromSq, toSq, Move.PR_QUEEN));
//        }
//    }

    public void makePC(int fromSq, long to){
        int toSq;
        while (to != 0){
            toSq = Bitboard.lsb(to);
            to = Bitboard.extractLsb(to);
            super.add(new Move(fromSq, toSq, Move.PC_KNIGHT));
            super.add(new Move(fromSq, toSq, Move.PC_BISHOP));
            super.add(new Move(fromSq, toSq, Move.PC_ROOK));
            super.add(new Move(fromSq, toSq, Move.PC_QUEEN));
        }
    }

    private void scoreMoves(final BoardState state, TranspositionTable transpositionTable) {
        final List<ScoredMove> result;

        if (this.size() == 0) {
            result = Collections.emptyList();
        } else {
            TTEntry ttEntry = transpositionTable.probe(state.hash());
            Move hashMove = ttEntry != null ? ttEntry.move() : null;

            result = this.stream().map(move -> {
                        int moveScore = move.equals(hashMove) ? MoveOrdering.HashMoveScore : 0;

//            if (moveOrdering.isKiller(state, move, ply)) {
//                move.addToScore(MoveOrdering.KillerMoveScore);
//            }
                        int piece = state.items[move.from()];
                        int piecesScore = switch (move.flags()) {
                            case Move.PC_BISHOP, Move.PC_KNIGHT, Move.PC_ROOK, Move.PC_QUEEN ->
                                    MGS[move.getPieceTypeForSide(state.getSideToPlay())][move.to()] - MGS[piece][move.from()]
                                            - MGS[state.items[move.to()]][move.to()];
                            case Move.PR_BISHOP, Move.PR_KNIGHT, Move.PR_ROOK, Move.PR_QUEEN ->
                                    MGS[move.getPieceTypeForSide(state.getSideToPlay())][move.to()] - MGS[piece][move.from()];
                            case Move.CAPTURE ->
                                    MGS[piece][move.to()] - MGS[piece][move.from()] - MGS[state.items[move.to()]][move.to()];
                            case Move.QUIET, Move.EN_PASSANT, Move.DOUBLE_PUSH, Move.OO, Move.OOO ->
                                    MGS[piece][move.to()] - MGS[piece][move.from()];
                            default -> throw new IllegalStateException();
                        };

                        int totalScore = moveScore + piecesScore * (state.getSideToPlay() == Side.WHITE ? 1 : -1);
                        return new ScoredMove(move, totalScore);
                    })
                    .collect(Collectors.toCollection(ArrayList::new));
        }

        this.sortedList = result;
    }

    public void pickNextBestMove(int curIndex){
        int max = Integer.MIN_VALUE;
        int maxIndex = -1;
        for (int i = curIndex; i < this.sortedList.size(); i++){
            if (this.sortedList.get(i).score > max){
                max = this.sortedList.get(i).score;
                maxIndex = i;
            }
        }
        Collections.swap(this.sortedList, curIndex, maxIndex);
    }

    // integrate scoreMoves with iterator
    public Iterable<Move> overSorted(final BoardState state, TranspositionTable transpositionTable/*, int ply, MoveOrdering moveOrdering*/) {
        scoreMoves(state, transpositionTable/*, ply, moveOrdering*/);
        // //        moves.scoreMoves(state, transpositionTable, 0, moveOrdering);
        ////        for (int i = 0; i < moves.size(); i++){
        ////            moves.pickNextBestMove(i);
        ////            Move move = moves.get(i);
        AtomicInteger index = new AtomicInteger();
        int count = this.sortedList.size();
        MoveList outer = this;
        return () -> new Iterator<>() {
            @Override
            public boolean hasNext() {
                return index.get() < count;
            }

            @Override
            public Move next() {
                int i = index.get();
                index.getAndIncrement();
                outer.pickNextBestMove(i);
                return outer.sortedList.get(i).move;
            }
        };
    }

    public record IndexedMove(Move move, int index) {}
}

File: ts-engine/src/MoveList.ts
import { Move } from './Move'; // Import Move and ScoredMove models
import { BoardState } from './BoardState';
import { Bitboard } from './Bitboard';
import { BB64Long } from './BB64Long';
import { Side } from './Side';
import {ScoredMove} from './ScoredMove';
import {TranspositionTable} from './TranspositionTable';
import {MoveOrdering} from './MoveOrdering';
import {PieceSquareTable} from './PieceSquareTable'; // Assuming Side is an enum for WHITE/BLACK

export class MoveList extends Array<Move> {

  constructor() {
    super();
  }

  makeQ(fromSq: number, to: BB64Long) {
    let toSq: number;
    while (!to.empty()) {
      toSq = to.LSB();
      to = to.popLSB();
      this.push(new Move(fromSq, toSq, Move.QUIET));
    }
  }

  makeC(fromSq: number, to: BB64Long) {
    let toSq: number;
    while (!to.empty()) {
      toSq = to.LSB();
      to = to.popLSB();
      this.push(new Move(fromSq, toSq, Move.CAPTURE));
    }
  }

  makeDP(fromSq: number, to: BB64Long) {
    let toSq: number;
    while (!to.empty()) {
      toSq = to.LSB();
      to = to.popLSB();
      this.push(new Move(fromSq, toSq, Move.DOUBLE_PUSH));
    }
  }

  makePC(fromSq: number, to: BB64Long) {
    let toSq: number;
    while (!to.empty()) {
      toSq = to.LSB();
      to = to.popLSB();
      this.push(new Move(fromSq, toSq, Move.PC_KNIGHT));
      this.push(new Move(fromSq, toSq, Move.PC_BISHOP));
      this.push(new Move(fromSq, toSq, Move.PC_ROOK));
      this.push(new Move(fromSq, toSq, Move.PC_QUEEN));
    }
  }

  private scoreMoves(state: BoardState, transpositionTable: TranspositionTable): ScoredMove[] {
    if (this.length === 0) {
      return [];
    }

    const ttEntry = transpositionTable.probe(state.hash);
    const hashMove: Move|null = ttEntry ? ttEntry.move() : null;

    return this.map((move: Move) => {
      let moveScore = hashMove !== null && hashMove.equals(hashMove) ? MoveOrdering.HashMoveScore : 0;

      const piece = state.items[move.from()];
      const piecesScore = (() => {
        switch (move.flags()) {
          case Move.PC_BISHOP:
          case Move.PC_KNIGHT:
          case Move.PC_ROOK:
          case Move.PC_QUEEN:
            return PieceSquareTable.MGS[move.getPieceTypeForSide(state.sideToPlay)][move.to()] -
              PieceSquareTable.MGS[piece][move.from()] - PieceSquareTable.MGS[state.items[move.to()]][move.to()];
          case Move.PR_BISHOP:
          case Move.PR_KNIGHT:
          case Move.PR_ROOK:
          case Move.PR_QUEEN:
            return PieceSquareTable.MGS[move.getPieceTypeForSide(state.sideToPlay)][move.to()] - PieceSquareTable.MGS[piece][move.from()];
          case Move.CAPTURE:
            return PieceSquareTable.MGS[piece][move.to()] - PieceSquareTable.MGS[piece][move.from()] - PieceSquareTable.MGS[state.items[move.to()]][move.to()];
          case Move.QUIET:
          case Move.EN_PASSANT:
          case Move.DOUBLE_PUSH:
          case Move.OO:
          case Move.OOO:
            return PieceSquareTable.MGS[piece][move.to()] - PieceSquareTable.MGS[piece][move.from()];
          default:
            throw new Error('Invalid move flag');
        }
      })();

      const totalScore = moveScore + piecesScore * (state.sideToPlay === Side.WHITE ? 1 : -1);
      return new ScoredMove(move, totalScore);
    });
  }

  pickNextBestMove(curIndex: number, sortedList: ScoredMove[]) {
    let max = Number.MIN_SAFE_INTEGER;
    let maxIndex = -1;
    for (let i = curIndex; i < sortedList.length; i++) {
      if (sortedList[i].score > max) {
        max = sortedList[i].score;
        maxIndex = i;
      }
    }
    if (maxIndex !== -1) {
      [sortedList[curIndex], sortedList[maxIndex]] = [sortedList[maxIndex], sortedList[curIndex]]; // Swap
    }
  }

  overSorted(state: BoardState, transpositionTable: TranspositionTable): IterableIterator<Move> {
    const sortedList = this.scoreMoves(state, transpositionTable);
    let index = 0;
    const moveList = this;

    return {
      [Symbol.iterator](): IterableIterator<Move> {
        return this;
      },

      next(): IteratorResult<Move> {
        if (index < sortedList.length) {
          moveList.pickNextBestMove(index, sortedList);
          const move = sortedList[index].move;
          // console.info(`returning idx: ${index}, move: ${move.toString()}, score: ${sortedList[index].score}`);
          index++;
          return { value: move, done: false };
        } else {
          return { value: undefined, done: true };
        }
      }
    };
  }
}
